### DISCUSSION TITLE
multiple hdfs via json

### DISCUSSION THREAD

**Discussion by wepl** [vAmigaWeb/vAmigaWeb#202] [Q&A]
Hi, is there a way start to start vAmigaWeb with multiple hdfs inserted (dh0: dh1: etc.)? 
I only saw 'url='.
And can provided hdfs be compressed in any way?

**Comment by mithrendal** [comment]
you have to provide a mount list like on the batman rises demo which mounts 2 floppies from an zipfile provided by json url param on the https://vamigaweb.github.io/doc/about.html page

here is the code snippet taken from the about page

```javascript
<a href="#" ontouchstart="touched=true" onclick="
let config={
	AROS:true,warpto:250,
	navbar:false,
	wide:true,
	gpu:true,
	display:'standard',
	startup_script: `df_mount_list=['Batman_Rises_disk1.adf','Batman_Rises_disk2.adf']`,
	url:'/doc/media/batman_rises.zip'								
}; 
vAmigaWeb_player.load(this,encodeURIComponent(JSON.stringify(config)));return false;" >		
	<img src="images/arrow.png" class="teaser_img" style="background-size: 100% 100%;background-image:url(images/batman_rises.jpg);">
</a>
```

in your case you would do to the same as in the batman rises demo but instead of `df_mount_list` you would use `hd_mount_list` in the startup_script... hope that works for you

**Reply by wepl** [reply]
Thanks for the advise. However I don't get it working at all also for a single hdf.
Firefox tells: Error: The operation is insecure.
Chrome: Error: Can't find end of central directory : is this a zip file ? If it is, see https://stuk.github.io/jszip/documentation/howto/read_zip.html
CORS are added to the HTTP headers.
Can you see the reason?
https://whdload.de/ctros/UnitA_FA18Interceptor_va.html

**Reply by mithrendal** [reply]
you are trying to load it from the vamigaweb site

see here your code

```javascript
	url:'/va/ctros/UnitA_FA18Interceptor.zip'
```

since the emulator is loading it, you have to point it to your site, see here 

```javascript
	  let config=
	  {
	    AROS:true,
	    navbar:false,
	    wide:true,
	    startup_script: `hd_mount_list=['UnitA_FA18Interceptor.hdf']`,
		url:'https://whdload.de/va/ctros/UnitA_FA18Interceptor.zip'
	  };
	  vAmigaWeb_player.load(this,encodeURIComponent(JSON.stringify(config)));
	  return false;
```

then it still complains because CORS headers are not set in your response

<img width="1090" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/13c41723-aeff-4014-982f-bca501a7d24e">

 looking in response header of whdload.de reveals CORS has not been set from yours site, see section "Antwortheader"

<img width="631" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/e577bb14-d91a-4694-bd0c-1be87b7a1136">


it needs to respond with headers 

`Access-Control-Allow-Origin: *`

or

`Access-Control-Allow-Origin: https://vAmigaWeb.github.io/`


besides this you are almost there 😎 when I disable CORS in the browser with a browser extension  it loads into the hdf


<img width="668" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/c402cbd0-cf91-491d-8b95-0aea61eec729">

**Comment by wepl** [comment]
Thanks it works.
I had set CORS header for the html page only, not for the zip file.
Maybe you could add to the wiki that the CORS HTTP header are relevant for data/images to download only.

**Comment by wepl** [comment]
Is there a way to avoid the requester 'insert in which dhx:'?
Is there a list of the possible options to set, e.g. cpu speed?

**Reply by mithrendal** [reply]
>I had set CORS header for the html page only, not for the zip file.

strange ... the CORS happens when vamigaweb fetches the zip from whdload ... so only that request is CORS relevant... I will check later. 

>Is there a way to avoid the requester 'insert in which dhx:'?

since you are providing the hd mount list it should not ask about insert in which dhx: ... the position in the mountlist should be enough... I will look into it after "the real work"

>Is there a list of the possible options to set, e.g. cpu speed?

it is possible but maybe not documented... I will give you a response later today

**Comment by mithrendal** [comment]
I saw that the zip file still had no CORS headers set, therefore it is still not working  ... better forget about CORS ... we better don't do CORS, it involves to much configuration on your server probably...

what we do now instead is to tell the vAmigaWeb_player to load the file. It is then no CORS (cross origin) anymore but same site (same origin), because the script was loaded by your page which belongs to your orgin. The player can handle that,  which is also described in the wiki...

do this

```javascript
vAmigaWeb_player.samesite_file=
{
  url:'/va/ctros/UnitA_FA18Interceptor.zip',
  name:'UnitA_FA18Interceptor.zip'
};

let config=
{
	AROS:true,
	navbar:false,
	wide:true,
	startup_script: `hd_mount_list=['UnitA_FA18Interceptor.hdf']`
};
vAmigaWeb_player.load(this,encodeURIComponent(JSON.stringify(config)));
return false;

```

>Is there a way to avoid the requester 'insert in which dhx:'?

regarding the prompt  ... avoiding it was only working for at least two mountable files e.g. hdfs or adfs ... 

Therefore (to make that it does not prompt for the drive number when only using one hdf) I just published a new version. now you can as in the script above, also use the mountlist with only one  mountable file. 

```javascript
hd_mount_list=['UnitA_FA18Interceptor.hdf'] //mounts it into dh0
hd_mount_list=[null, 'UnitA_FA18Interceptor.hdf'] //mounts it into dh1
hd_mount_list=[null, null, 'UnitA_FA18Interceptor.hdf'] //mounts it into dh2
//and so on
```
 
I will look into your other questions tomorrow...

**Reply by wepl** [reply]
Thanks a lot. That works.

**Reply by mithrendal** [reply]
>Thanks a lot. That works.

you are welcome 😎

>Is there a list of the possible options to set, e.g. cpu speed?

just append commands for vm machine configuration to the `startup_script` 

e.g.

```javascript
//clock rate == 21 Mhz == 3 * 7Mhz, reasonable up to 14 which is 100Mhz
wasm_configure('CPU_OVERCLOCKING', '3');

//68020=2, 68010=1, 68000=0
wasm_configure('CPU_REVISION', '2');

//maximum floppy drive speed
wasm_configure('DRIVE_SPEED', '8');

//agnus options 'OCS_OLD','OCS','ECS_1MB','ECS_2MB'
wasm_configure('AGNUS_REVISION', 'ECS_2MB');

//denise options 'OCS','ECS'
wasm_configure('DENISE_REVISION', 'OCS');

//chip ram options '256', '512', '1024', '2048'
wasm_configure('CHIP_RAM', '512');

//slow ram options '0', '256', '512'
wasm_configure('SLOW_RAM', '512');

//fast ram options '0', '256', '512','1024', '2048', '8192'
wasm_configure('FAST_RAM', '8192');

//ntsc or pal machine 'pal', 'ntsc'
wasm_set_display('pal');
```


just add these configuration commands to the end of startup_script , 
e.g. 68020 at 28 Mhz, with 8MB of fast ram

```javascript
startup_script: `
hd_mount_list=['UnitA_FA18Interceptor.hdf'];
wasm_configure('CPU_REVISION', '2');
wasm_configure('CPU_OVERCLOCKING', '4');
wasm_configure('FAST_RAM', '8192');
`
```

**Reply by wepl** [reply]
Thanks a lot.
I have added vAmigaWeb for all Cracktros now: https://whdload.de/ctros/alla.html
At least the ones which not depend on kickemu13.
It's planned to do this for all Mags and Demos too.

**Reply by mithrendal** [reply]
cool 😎

to inject other kickstart into vAmigWeb you would have to add the kickstart_rom_url 


```javascript
vAmigaWeb_player.samesite_file=
{
  kickstart_rom_url: '/va/roms/kick2.rom',
  url:'/va/ctros/UnitA_FA18Interceptor.zip',
  name:'UnitA_FA18Interceptor.zip'
};

```
and to instruct it to NOT use AROS and instead wait for the kick rom loading to complete

```javascript
let config=
{
	AROS:false,
        wait_for_kickstart_injection:true,
...

```

**Reply by wepl** [reply]
The problem is that I cannot provide a kickstart for copyright reasons.
It would also be needed inside the hdf.
Maybe it is possible to let the user provide a hdf which contains it. However, I would leave this for later.

**Comment by wepl** [comment]
> to inject other kickstart into vAmigWeb you would have to add the kickstart_rom_url

how would that look if I like to inject also the extended rom for AROS?

**Comment by mithrendal** [comment]
>how would that look if I like to inject also the extended rom for AROS?

oops... currently not implemented ;-) ... will do that tonight if you need this. just tell me

I would propose to add a `kickstart_ext_url` property to the file descriptor like you see below 


```javascript
vAmigaWeb_player.samesite_file=
{
  kickstart_rom_url: '/va/roms/kickAros.rom',
  kickstart_ext_url: '/va/roms/kickArosExt.rom',
  url:'/va/ctros/UnitA_FA18Interceptor.zip',
  name:'UnitA_FA18Interceptor.zip'
};
```

**Reply by wepl** [reply]
The propose is fine. But no hurry, ATM I have no ROM found which works properly.

**Reply by mithrendal** [reply]
I just pushed a new version which supports `kickstart_ext_url` property in the file descriptor

but I did not changed the version number of the web app (in order to not annoy the other guys out there with our tests 😬 and showing the upgrade dialog), so to try out just use private browsing (so that it fetches it new from the server) or just delete the cache so that it loads the upgraded version ...

in the settings 

<img width="848" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/215b8ce0-f168-4d07-a7ce-ec59c31a2282">

click on the red trashcan deletes the currently selected version from cache

after you deleted the version you can reinstall with

<img width="848" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/9c5f3aa1-62cc-43de-879f-1e6f7b0e0bfc">

a click on the green button

**Reply by wepl** [reply]
Thanks, but it doesn't seem to load the kickstarts: https://whdload.de/ctros/UnitA_FA18InterceptorX.html

**Reply by wepl** [reply]
must it be named .rom?

**Reply by mithrendal** [reply]
Wait I will check

**Reply by mithrendal** [reply]
copy&paste error🙈 ... I put the kick rom in both rom slots .... instead of the ext file...

>must it be named .rom?

no it is name agnostic... you can name them as you like

I just pushed the fixed version

seems to work fine now, when agnus is manually set to OCS

<img width="900" alt="image" src="https://github.com/vAmigaWeb/vAmigaWeb/assets/17108995/17909014-bb5c-44b0-99e1-c844298e0d69">



EDIT: oh you turned the warp on .... it warps like crazy now ... always 😂

**Reply by wepl** [reply]
Great, thanks.
Yes, works fine with OCS Agnus. But 512k Chip is limiting.

**Comment by wepl** [comment]
> just append commands for vm machine configuration to the `startup_script`
> 
> e.g.
> 
> ```js
> //clock rate == 21 Mhz == 3 * 7Mhz, reasonable up to 14 which is 100Mhz
> wasm_configure('CPU_OVERCLOCKING', '3');
> 
> //68020=2, 68010=1, 68000=0
> wasm_configure('CPU_REVISION', '2');
> 
> //maximum floppy drive speed
> wasm_configure('DRIVE_SPEED', '8');
> 
> //agnus options 'OCS_OLD','OCS','ECS_1MB','ECS_2MB'
> wasm_configure('AGNUS_REVISION', 'ECS_2MB');
> 
> //denise options 'OCS','ECS'
> wasm_configure('DENISE_REVISION', 'OCS');
> 
> //chip ram options '256', '512', '1024', '2048'
> wasm_configure('CHIP_RAM', '512');
> 
> //slow ram options '0', '256', '512'
> wasm_configure('SLOW_RAM', '512');
> 
> //fast ram options '0', '256', '512','1024', '2048', '8192'
> wasm_configure('FAST_RAM', '8192');
> ```
> 
> just add these configuration commands to the end of startup_script , e.g. 68020 at 28 Mhz, with 8MB of fast ram
> 
> ```js
> startup_script: `
> hd_mount_list=['UnitA_FA18Interceptor.hdf'];
> wasm_configure('CPU_REVISION', '2');
> wasm_configure('CPU_OVERCLOCKING', '4');
> wasm_configure('FAST_RAM', '8192');
> `
> ```

It looks like this doesn't work. Used for example here https://whdload.de/ctros/Classic_ManchesterUtdEur.html
But when I check the running emu the values are not set.
I'm doing something wrong?

**Reply by mithrendal** [reply]
I think it works internally but is not shown in the emu settings... 

the emu thinks that can only be changed by himself in the settings... so it does not refresh the values when settings dialog is opened.

I will check

**Reply by mithrendal** [reply]
confirmed it does not work.  I am on it ... hang on

**Reply by mithrendal** [reply]
Ok I see the error ...   you have to put the wasm_configure commands all into the `on_ready_to_run` hook 😎

**Comment by wepl** [comment]
Ok, I have changed it :-)

**Comment by wepl** [comment]
Is there also an option for the amount of floppy drives?
I don't need second drive and it eats memory.

**Reply by mithrendal** [reply]
```javascript
wasm_configure('floppy_drive_count','1');
```

just add that to the other configs in the `on_ready_to_run` 😎

'1' also shows no drive prompt when dragging files manually into vAmigaWeb, since there is nothing to choose with only 1 option 😀 


 >it eats memory.

interesting stuff 🤓, how much does it eat ?

**Reply by wepl** [reply]
thanks, works.
cbm: 16k chip
aros: 24k fast
